### DISCUSSION TITLE
InkAnnotations, writing, and erasing.

### DISCUSSION THREAD

**Discussion by legraina** [stephanrauh/ngx-extended-pdf-viewer#2480] [General]
Hi,

I'm trying to build an app where we can write with a pencil on some pdf.
However, while the rest of the library is excellent, I have many challenges creating a good user experience, especially the possibility of erasing some writing with an eraser tool like many tablets' apps could propose.
I was wondering if it's a project on which you are working or the general architecture of the library is not designed for this type of usage ?
Thanks a lot,

Antoine

**Comment by stephanrauh** [comment]
@calixteman

**Comment by legraina** [comment]
Hi @stephanrauh,

I've been able to define my own eraser and improve the writing experience by playing around the DOM. Would you like me to share my angular component ?
Also, I'm trying to be able to use the webapp I have developed offline (or at least without internet for a while), but I have encountered an issue: the pdf.worker is reloaded everytime I'm changing the pdf source. I thought that by keeping the same component it would not behave this way. I was wondering if I was doing something badly or if it's the normal behavior ? (I can create another thread if you like)
Thanks

**Reply by stephanrauh** [reply]
Sharing is always nice!

As for the worker: yes, creating a fresh worker every time is the intended behaviour, or at least it's the same behavior as https://mozilla.github.io/pdf.js/web/viewer.html shows. But I also wondered about this. It seems to be possible to create the worker outside pdf.js and to reuse it (see https://github.com/mozilla/pdf.js/issues/7057). Let's open a dedicated ticket for this.

**Comment by legraina** [comment]
No problem. 

You can find in the zip the component I've created to write on a tablet with a good ux. To do so, I had to disable the resizers and the possibility to delete layers. I have also implemented an observer to automatically unselect selected layers. For the eraser option, I had to delete the right points of the annotations, and then remove and re-add the annotations and rebuilt the bezier points (here, there is a small issue when creating the modified annotations, there is a very small translation of a few pixels due to the reconstruction of the bezier curve).

<img width="973" alt="image" src="https://github.com/user-attachments/assets/34709cc8-565c-4204-9782-2a97997368b1">

[pdf-viewer.component.zip](https://github.com/user-attachments/files/16620471/pdf-viewer.component.zip)

**Reply by stephanrauh** [reply]
C'etait beaucoup de travail! I'm afraid it's going to take a while, but I'd like to add your component to the showcase. Among other things, it's a very nice example of adding custom functionality to the toolbar. Keep on the good work!
